src/actions/loadData.js

```diff
@@ -263,104 +198,225 @@ const narrativeFetchingErrorNotification = (err, failedTreeName, fallbackTreeNam
+    dispatch(goTo404(msg));
+  }
+}
+
+/**
+ * @returns {Promise}
+ */
+function fetchAndParseNarrative(url) {
+  const apiCall = `${getServerAddress()}/getNarrative?prefix=${url}&type=md`;
+  return fetchWithErrorHandling(apiCall)
+    .then((res) => res.text())
+    .then((res) => parseMarkdownNarrativeFile(res, parseMarkdown))
+    .catch((err) => {
+      /* Note that up until v2.16 the client expected the narrative to have been converted
       to JSON by the server. To facilitate backward compatibility (e.g. in the case where
       the client is >2.16, but the server is using an older version of the API)
-      if the file doesn't look like markdown we will attempt to parse it as JSON */
-      getDatasetFromCharon(url, {narrative: true, type: "md"})
-        .then((res) => res.text())
-        .then((res) => parseMarkdownNarrativeFile(res, parseMarkdown))
-        .catch((err) => {
-          // errors from `parseMarkdownNarrativeFile` indicating that the file doesn't look
-          // like markdown will have the fileContents attached to them
-          if (!err.fileContents) throw err;
-          console.error("Narrative file doesn't appear to be markdown! Attempting to parse as JSON.");
-          return JSON.parse(err.fileContents);
-        })
-        .then((blocks) => {
-          addEndOfNarrativeBlock(blocks);
-          return fetchAndCacheNarrativeDatasets(dispatch, blocks, query);
-        })
-        .catch((err) => {
-          console.error("Error obtaining narratives", err.message);
-          dispatch(goTo404(`Couldn't load narrative for ${url}`));
-        });
+      if the file doesn't look like markdown we will attempt to parse it as JSON.
+      Errors from `parseMarkdownNarrativeFile` indicating that the file doesn't look
+      like markdown will have the fileContents attached to them */
+      if (!err.fileContents) throw err;
+      console.error("Narrative file doesn't appear to be markdown! Attempting to parse as JSON.");
+      return JSON.parse(err.fileContents);
+    })
+    .then((blocks) => {
+      addEndOfNarrativeBlock(blocks);
+      return blocks;
+    });
+}
+
+
+function addEndOfNarrativeBlock(narrativeBlocks) {
+  const lastContentSlide = narrativeBlocks[narrativeBlocks.length-1];
+  const endOfNarrativeSlide = Object.assign({}, lastContentSlide, {
+    __html: undefined,
+    isEndOfNarrativeSlide: true
+  });
+  narrativeBlocks.push(endOfNarrativeSlide);
+}
+
+/**
+ * A narrative (parsed into `blocks`) may define multiple datasets.
+ * This function returns the API endpoints needed to request these
+ * datasets (including sidecar files).
+ * Note: "initial": starting page, potentially different to frontmatter
+ */
+function createApiCallsForDatasetsInNarratives(blocks, query) {
+  let initialDatasetNames, frontmatterDatasetNames;
+  const datasetApiCalls = {};
+  const initialBlockIdx = getNarrativePageFromQuery(query, blocks);
+  collectDatasetFetchUrls(blocks[initialBlockIdx].dataset);
```

Comment:
I'm not sure I understand what calling this here does - I know it modifies the dataset url, but won't it also get called in the loop below in the same way? Does it need to happen twice for `blocks[initialBlockIdx].dataset`?

Reply:
Code updated, so I think this can be resolved?